### ISSUE TITLE
[RFC] merge offcputime and offwaketime

### ISSUE THREAD

**Issue by abirchall** [iovisor/bcc#545]
I recently made a couple of changes to offcputime.py (#533, #543 and #544). I was going to port them to offwaketime.py when it occurred to me that including the stack of the waking process could be a feature of offcputime.py (that was default off). I'm envisioning something like `parser.add_argument('-w', '--show-wakeup-stack', ...)`.
How do we feel about merging these two tools into one vs maintaining separate tools?
I can take this on if we think it's a good idea.

cc @brendangregg

**Comment by brendangregg** [comment]
I'm a bit nervous about making scripts too big, as they become harder to learn, maintain, and test. Especially since they are kprobe based, which makes them more brittle in theory, and they should be tracepoint based (at which point, in theory, there's less need to go maintain and test them!).

But yes, ultimately they could be merged as you said, and I wouldn't oppose doing it, since it seems to make sense. Just wouldn't go too far with combining other tools: the Unix model of small tools that do one thing well, works well.

Probably more important right now is to use a stack trace map for offwaketime.py (needs Linux 4.6?), ie, cp offwaketime.py to tools/old, then update offwaketime.py to use the new map feature. The changes should be very similar to what Vicent did for offcputime: see https://github.com/iovisor/bcc/commit/e82fb1baa26b0e1e524923765ba70d5872e51eef

Then things get really exciting, since it'll get easier to implement chain graphs:

http://www.brendangregg.com/blog/2016-02-05/ebpf-chaingraph-prototype.html

Although how to actually do chain graphs well I think is an unsolved problem, as described in that post in "current issues", but then I wrote that before the stack maps existed, so things should be much easier ow.

**Comment by abirchall** [comment]
> Probably more important right now is to use a stack trace map for offwaketime.py (needs Linux 4.6?), ie, cp offwaketime.py to tools/old, then update offwaketime.py to use the new map feature. The changes should be very similar to what Vicent did for offcputime: see e82fb1b

Agree. This was part of my motivation for creating this RFC; that combining the two scripts may be more practical given the changes that are required to bring offwaketime up to date.

> I'm a bit nervous about making scripts too big, as they become harder to learn, maintain, and test. Especially since they are kprobe based, which makes them more brittle in theory, and they should be tracepoint based (at which point, in theory, there's less need to go maintain and test them!).

Agree. I feel like there's a trade off here. On one hand, in adding more options to a single tool we add complexity to the tool as a whole (especially given that we're generating code on the fly based on these arguments). On the other hand there's also a higher maintenance cost in having distinct tools that take similar actions, that have similar features and that need to handle the same edge cases.

How about I take a stab at this and we can re-evaluate when we see how complex the merged tool turns out?

**Comment by 4ast** [comment]
I think Brendan was objecting too merging of tools earlier also due to human factor of 'unix model'. I think if offcputime and offwaketime become one script, it would make sense to make it hard-linkable, so it changes behavior depending on the name of the link, then we'll have good code reuse, keep unix philosophy and users won't be surprised. Thoughts?

**Comment by brendangregg** [comment]
The hardlinks idea was not just about small tools, but also about discoverability when searching for and browsing tools. It lets you keep the keywords in the tool names: in that case, keywords like "ext4", "btrfs", "xfs", "zfs" that people would be looking for.

I don't think anyone is looking for "offwake" yet as a keyword, so having it show up in a listing for discoverability probably isn't important. So I think I'd be ok with having it as an option to offcputime.

Chain graphs, on the other hand, may well be a "thing", and we may want a "chaingraph.py" tool (or hardlink) that people can easily find. Like how people currently go hunting for flame graphs.

**Comment by 4ast** [comment]
I mean both offwaketime and offcputime will exist in an install directory as separate files (and discoverability will be preserved) by they will point to the same .py or .lua script. Either as hard link, soft link or two copies of the same file. The script will take the default action based on the name of the file.
I think *fsdict are primary candidates for such code sharing. If we would need to un-merge them later due to diverging implementation we can certainly do that. Right now it seems the code is quite similar.

**Comment by brendangregg** [comment]
Yes, they can be merged (soft links). I just think offwaketime might actually be ok to become a switch to offcputime. I'm happy either way.

The offcputime program is a great template; would be nice to have a stripped down version in /examples.

... Just reminds me that I'd also like offcpuhist and cpuhist (or runhist) tools to show me histograms of off-CPU time and on-CPU respectively. Model them on runqlat.
